pkg/distributor/distributor.go

```diff
@@ -1846,6 +1846,45 @@ func (d *Distributor) ActiveSeries(ctx context.Context, matchers []*labels.Match
 	return deduplicatedSeries, nil
 }
 
+// activeSeriesResponse is a helper to merge/deduplicate ActiveSeries responses from ingesters.
+type activeSeriesResponse struct {
+	m       sync.Mutex
+	series  map[uint64]labels.Labels
+	builder labels.ScratchBuilder
+	lbls    labels.Labels
+}
+
+func newActiveSeriesResponse() *activeSeriesResponse {
+	return &activeSeriesResponse{
+		series:  make(map[uint64]labels.Labels),
+		builder: labels.NewScratchBuilder(40),
+	}
+}
+
+func (r *activeSeriesResponse) add(series []*mimirpb.Metric) {
+	r.m.Lock()
+	defer r.m.Unlock()
+
+	for _, metric := range series {
+		mimirpb.FromLabelAdaptersOverwriteLabels(&r.builder, metric.Labels, &r.lbls)
```

Comment:
sorry, this is probably a silly question, what is the point of `r.builder` and of using this function if the function looks like this:

```
// This is like FromLabelAdaptersToLabels but easier for stringlabels to implement.
func FromLabelAdaptersOverwriteLabels(_ *labels.ScratchBuilder, ls []LabelAdapter, dest *labels.Labels) {
	*dest = FromLabelAdaptersToLabels(ls)
}
```

why not

```suggestion
		r.lbls = mimirpb.FromLabelAdaptersToLabels(metric.Labels)
```

Reply:
The implementation you're referencing is behind `//go:build !stringlabels`. When built with `stringlabels`, it will instead do this
```
// Build a labels.Labels from LabelAdaptors, with amortized zero allocations.
func FromLabelAdaptersOverwriteLabels(builder *labels.ScratchBuilder, ls []LabelAdapter, dest *labels.Labels) {
	builder.Reset()
	for _, v := range ls {
		builder.Add(v.Name, v.Value)
	}
	builder.Overwrite(dest)
}
```